### PR TITLE
update Beats generator

### PR DESCRIPTION
**Pull request by cyrilleverrier**
This code generates a code similar to https://github.com/elastic/apm-server

- build scripts are copied to `_beat` folder
- dependencies are vendored using `govendor` in the `vendor` folder

### CODE REVIEW

**Comment by cyrilleverrier** [comment]
the tests are not yet updated because using govendor to copy the dependencies takes a lot of time.
I didn't want to break your build system

Please let me know if this PR is useful. If so, the same should be done for the metricbeat generator

File: dev-tools/generate_notice.py [hunk 1]
```diff
@@ -242,6 +242,8 @@ def detect_license_summary(content):
                         help="Beat name")
     parser.add_argument("-c", "--copyright", default="Elasticsearch BV",
                         help="copyright owner")
+    parser.add_argument("-y", "--copyright-year", default="2014",
```
Comment (ruflin) [inline dev-tools/generate_notice.py]: We should set the default probably to 2018.

File: generator/beat/{beat}/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
+FROM golang:1.9.2
```
Comment (ruflin) [inline generator/beat/{beat}/Dockerfile]: Nice, even the docker file.

File: generator/beat/{beat}/Makefile [hunk 3]
```diff
@@ -2,47 +2,78 @@ BEAT_NAME={beat}
...
-NOTICE_FILE=NOTICE
-
-# Path to the libbeat Makefile
--include $(ES_BEATS)/libbeat/scripts/Makefile
+BEAT_DOC_URL=https://path/to/{beat}/doc
+BEAT_DESCRIPTION=My description
+BEAT_LICENSE=My license
+COPYRIGHT_TXT=My company
+COPYRIGHT_YEAR={current_year}
+VERSION=0.1.0
```
Comment (ruflin) [inline generator/beat/{beat}/Makefile]: Should this be by default the same version as Beats?

File: generator/beat/{beat}/Makefile [hunk 4]
```diff
@@ -2,47 +2,78 @@ BEAT_NAME={beat}
...
+	govendor fetch github.com/elastic/beats/libbeat/kibana/@$(BEATS_VERSION)
+	scripts/patch_vendorjson.py
+	BEATS_VERSION=$(BEATS_VERSION) scripts/update_beats.sh
+	scripts/patch_vendorjson.py --done
+	BEATS_VERSION=$(BEATS_VERSION) scripts/update_beats.sh
+	$(MAKE) update
+	@echo --- Use this commit message: Update beats framework to `cat vendor/vendor.json | python -c 'import sys, json; print([p["revision"] for p in json.load(sys.stdin)["package"] if p["path"] == "github.com/elastic/beats/libbeat/beat"][0][:7])'`
+
+update-{beat}-dependencies:
+	govendor fetch github.com/olivere/elastic@v6.1.2
```
Comment (ruflin) [inline generator/beat/{beat}/Makefile]: ?

File: generator/beat/{beat}/scripts/clean_vendor.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
Comment (ruflin) [inline generator/beat/{beat}/scripts/clean_vendor.sh]: I think this one is the same (or almost the same) file we have in the top level beats/script directory. I wonder if we could copy this over instead?

File: generator/beat/{beat}/scripts/patch_vendorjson.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+
+
+"""
+Update dependencies of the given govendor'd package.  For https://github.com/elastic/apm-server/issues/450.
```
Comment (ruflin) [inline generator/beat/{beat}/scripts/patch_vendorjson.py]: Probably no `apm-server` should show up here.

File: testing/environments/docker/kibana/Dockerfile [hunk 7]
```diff
@@ -8,7 +8,7 @@ EXPOSE 5601
 ARG DOWNLOAD_URL=https://snapshots.elastic.co/downloads
 ARG ELASTIC_VERSION=7.0.0-alpha1-SNAPSHOT
 ARG CACHE_BUST=1
-ARG XPACK=1
```
Comment (ruflin) [inline testing/environments/docker/kibana/Dockerfile]: Is this change intentional. I think this will affect our default testing.

File: generator/beat/{beat}/Makefile (hunk 3 above)
Comment (cyrilleverrier) [reply generator/beat/{beat}/Makefile]: Up-to-you to decide. But my own preference is  to have a different version number between the application and the libbeat library.

File: testing/environments/docker/kibana/Dockerfile (hunk 7 above)
Comment (cyrilleverrier) [reply testing/environments/docker/kibana/Dockerfile]: see `testing/environments/docker/elasticsearch/Dockerfile`
```
ARG ELASTIC_VERSION
ARG DOWNLOAD_URL
ARG ES_JAVA_OPTS
ARG CACHE_BUST=1
ARG XPACK
```

The XPACK value for Elasticsearch and Kibana should be set by `testing/environments/snapshot.yml` and `testing/environments/snapshot-noxpack.yml`  


Shall I create a separate PR for this issue?

Comment (cyrilleverrier) [reply testing/environments/docker/kibana/Dockerfile]: I created https://github.com/elastic/beats/pull/6383/

**Comment by ruflin** [comment]
@cyrilleverrier Sorry, didn't get to this yet but definitively on my list. Could you rebase again on master?

File: dev-tools/set_version [hunk 8]
```diff
@@ -5,16 +5,15 @@ import re
 import sys
 from subprocess import check_call
 
-vendored_libbeat = os.path.normpath("vendor/github.com/elastic/beats")
+vendored_libbeat = os.path.normpath("_beat")
 
 
-goversion_template = '''package main
+goversion_template = '''package cmd
```
Comment (ruflin) [inline dev-tools/set_version]: Is this in `cmd`? Just double checking.

File: generator/beat/{beat}/Makefile [hunk 9]
```diff
@@ -2,47 +2,84 @@ BEAT_NAME={beat}
...
-	mkdir -p vendor/github.com/elastic/
-	cp -R ${BEAT_GOPATH}/src/github.com/elastic/beats vendor/github.com/elastic/
-	rm -rf vendor/github.com/elastic/beats/.git
+SYSTEM_TESTS=true
+TEST_ENVIRONMENT=true
+TESTING_ENVIRONMENT=snapshot-noxpack
 
-.PHONY: git-init
-git-init:
+# Path to the libbeachmt Makefile
```
Comment (ruflin) [inline generator/beat/{beat}/Makefile]: s/libbeachmt/libbeat ?

File: generator/beat/{beat}/Makefile [hunk 10]
```diff
@@ -2,47 +2,84 @@ BEAT_NAME={beat}
...
+vendor:
+	mkdir -p vendor
+	govendor init
+
+update-dependencies: vendor
+	$(MAKE) update-libbeat
+	$(MAKE) update-{beat}-dependencies
+
+update-libbeat:
+	@chmod +x scripts/clean_vendor.sh
```
Comment (ruflin) [inline generator/beat/{beat}/Makefile]: It's pretty hard to review in which case it's going to work and when not. Best lets get this in an play with it :-)

File: generator/beat/{beat}/Makefile [hunk 11]
```diff
@@ -2,47 +2,84 @@ BEAT_NAME={beat}
...
 # This is called by the beats packer before building starts
 .PHONY: before-build
 before-build:
 
 # Collects all dependencies and then calls update
 .PHONY: collect
-collect:
+collect: notice
+
+check-full: check
```
Comment (ruflin) [inline generator/beat/{beat}/Makefile]: Add `.PHONY: check-full`

File: generator/beat/{beat}/README.md [hunk 12]
```diff
@@ -9,7 +9,7 @@ Ensure that this folder is at the following location:
 
 ### Requirements
 
-* [Golang](https://golang.org/dl/) 1.7
+* [Golang](https://golang.org/dl/) 1.9.2
```
Comment (ruflin) [inline generator/beat/{beat}/README.md]: Would be great to have this as a variable in the future (not for now)

**Comment by cyrilleverrier** [comment]
Current issues when running `make test` in the generator folder:

1. the generated testbeat's Makefile calls `govendor fetch github.com/elastic/beats/...@master`. Therefore, the test is slow to run and testbeat is compiled against libbeat from the master branch instead of libbeat from the current commit

2. metricbeats generator not yet modified to use govendor. Test fails.

**Comment by cyrilleverrier** [comment]
Code has been rebased on top of master (768e58ede27f7a80b38c983a83537f8669a0cb04)

**Comment by ruflin** [comment]
@cyrilleverrier Thanks, will try to find some time this week to have a look at the metricbeat generator, especially why one relies on the other.
